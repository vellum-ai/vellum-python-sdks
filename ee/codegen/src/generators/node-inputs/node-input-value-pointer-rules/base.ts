import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";

import { WorkflowContext } from "src/context";
import { BaseNodeContext } from "src/context/node-context/base";
import {
  IterableConfig,
  NodeInputValuePointerRule as NodeInputValuePointerRuleType,
  WorkflowDataNode,
} from "src/types/vellum";

export declare namespace BaseNodeInputValuePointerRule {
  export interface Args<T extends NodeInputValuePointerRuleType> {
    nodeContext: BaseNodeContext<WorkflowDataNode>;
    nodeInputValuePointerRule: T;
    iterableConfig?: IterableConfig;
  }
}

export abstract class BaseNodeInputValuePointerRule<
  T extends NodeInputValuePointerRuleType
> extends AstNode {
  private readonly nodeContext: BaseNodeContext<WorkflowDataNode>;
  public readonly workflowContext: WorkflowContext;
  public readonly nodeInputValuePointerRule: T;
  public readonly iterableConfig?: IterableConfig;
  private astNode: AstNode | undefined;

  constructor({
    nodeContext,
    nodeInputValuePointerRule,
    iterableConfig,
  }: BaseNodeInputValuePointerRule.Args<T>) {
    super();
    this.nodeContext = nodeContext;
    this.workflowContext = nodeContext.workflowContext;
    this.iterableConfig = iterableConfig;
    this.nodeInputValuePointerRule = nodeInputValuePointerRule;

    this.astNode = this.getAstNode();
    if (this.astNode) {
      this.inheritReferences(this.astNode);
    }
  }

  public getReferencedNodeContext():
    | BaseNodeContext<WorkflowDataNode>
    | undefined {
    return undefined;
  }

  abstract getAstNode(): AstNode | undefined;

  public write(writer: Writer): void {
    if (this.astNode) {
      this.astNode.write(writer);
    }
  }
}
