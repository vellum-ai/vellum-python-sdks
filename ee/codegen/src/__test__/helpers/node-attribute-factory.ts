import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
