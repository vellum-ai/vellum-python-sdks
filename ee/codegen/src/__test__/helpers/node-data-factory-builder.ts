import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
