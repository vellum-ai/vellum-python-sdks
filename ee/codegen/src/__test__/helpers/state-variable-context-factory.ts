import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
