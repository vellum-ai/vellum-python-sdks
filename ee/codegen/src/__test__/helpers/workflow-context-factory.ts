import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
