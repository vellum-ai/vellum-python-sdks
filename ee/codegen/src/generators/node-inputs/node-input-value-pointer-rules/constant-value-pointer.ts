import { BaseNodeInputValuePointerRule } from "./base";

import * as codegen from "src/codegen";
import { BaseNodeContext } from "src/context/node-context/base";
import { VellumValue } from "src/generators";
import {
  ConstantValuePointer,
  IterableConfig,
  WorkflowDataNode,
} from "src/types/vellum";

export declare namespace ConstantValuePointerRule {
  interface Args {
    nodeContext: BaseNodeContext<WorkflowDataNode>;
    nodeInputValuePointerRule: ConstantValuePointer;
    iterableConfig?: IterableConfig;
  }
}

export class ConstantValuePointerRule extends BaseNodeInputValuePointerRule<ConstantValuePointer> {
  constructor(args: ConstantValuePointerRule.Args) {
    super(args);
  }

  getAstNode(): VellumValue {
    const constantValuePointerRuleData = this.nodeInputValuePointerRule.data;

    return codegen.vellumValue({
      vellumValue: constantValuePointerRuleData,
      iterableConfig: this.iterableConfig,
    });
  }
}
