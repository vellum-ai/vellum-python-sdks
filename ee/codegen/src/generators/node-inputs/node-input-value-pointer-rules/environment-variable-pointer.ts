import { python } from "@fern-api/python-ast";

import { BaseNodeInputValuePointerRule } from "./base";

import { EnvironmentVariablePointer } from "src/types/vellum";

export class EnvironmentVariablePointerRule extends BaseNodeInputValuePointerRule<EnvironmentVariablePointer> {
  getAstNode(): python.AstNode {
    const envVarName = this.nodeInputValuePointerRule.data.environmentVariable;
    return python.instantiateClass({
      classReference: python.reference({
        name: "EnvironmentVariableReference",
        modulePath: [
          ...this.workflowContext.sdkModulePathNames.WORKFLOWS_MODULE_PATH,
          "references",
        ],
      }),
      arguments_: [
        python.methodArgument({
          value: python.TypeInstantiation.str(envVarName),
        }),
      ],
    });
  }
}
