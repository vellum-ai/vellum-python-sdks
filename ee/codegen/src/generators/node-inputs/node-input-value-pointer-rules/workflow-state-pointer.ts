import { python } from "@fern-api/python-ast";
import { AstNode } from "@fern-api/python-ast/core/AstNode";

import { BaseNodeInputValuePointerRule } from "./base";

import { BaseNodeContext } from "src/context/node-context/base";
import { NodeInputNotFoundError } from "src/generators/errors";
import { WorkflowStatePointer, WorkflowDataNode } from "src/types/vellum";

export declare namespace WorkflowStatePointerRule {
  interface Args {
    nodeContext: BaseNodeContext<WorkflowDataNode>;
    nodeInputValuePointerRule: WorkflowStatePointer;
  }
}

export class WorkflowStatePointerRule extends BaseNodeInputValuePointerRule<WorkflowStatePointer> {
  constructor(args: WorkflowStatePointerRule.Args) {
    super(args);
  }

  getAstNode(): AstNode {
    const workflowStatePointerRuleData = this.nodeInputValuePointerRule.data;

    const stateVariableContext =
      this.workflowContext.findStateVariableContextById(
        workflowStatePointerRuleData.stateVariableId
      );

    if (!stateVariableContext) {
      this.workflowContext.addError(
        new NodeInputNotFoundError(
          `Could not find state variable context with id ${workflowStatePointerRuleData.stateVariableId}`,
          "WARNING"
        )
      );
      return python.TypeInstantiation.none();
    }

    return python.reference({
      name: stateVariableContext.definition.name,
      modulePath: stateVariableContext.definition.module,
      attribute: [stateVariableContext.name],
    });
  }
}
