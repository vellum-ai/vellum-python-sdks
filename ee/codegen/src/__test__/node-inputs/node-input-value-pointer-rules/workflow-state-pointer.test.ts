import { Writer } from "@fern-api/python-ast/core/Writer";

import {
  nodeContextFactory,
  workflowContextFactory,
} from "src/__test__/helpers";
import { stateVariableContextFactory } from "src/__test__/helpers/state-variable-context-factory";
import { WorkflowStatePointerRule } from "src/generators/node-inputs/node-input-value-pointer-rules/workflow-state-pointer";

describe("WorkflowStatePointer", () => {
  let writer: Writer;

  beforeEach(() => {
    writer = new Writer();
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("should generate correct Python code", async () => {
    const workflowContext = workflowContextFactory();
    workflowContext.addStateVariableContext(
      stateVariableContextFactory({
        stateVariableData: {
          id: "test-state-id",
          key: "state_1",
          type: "STRING",
          default: null,
          required: false,
          extensions: { color: "cyan" },
        },
        workflowContext,
      })
    );

    const nodeContext = await nodeContextFactory({ workflowContext });

    const workflowStatePointer = new WorkflowStatePointerRule({
      nodeContext,
      nodeInputValuePointerRule: {
        type: "WORKFLOW_STATE",
        data: {
          stateVariableId: "test-state-id",
        },
      },
    });

    workflowStatePointer.write(writer);

    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should handle when it's referencing a state variable that no longer exists", async () => {
    const workflowContext = workflowContextFactory({ strict: false });
    const nodeContext = await nodeContextFactory({ workflowContext });

    const workflowStatePointer = new WorkflowStatePointerRule({
      nodeContext,
      nodeInputValuePointerRule: {
        type: "WORKFLOW_STATE",
        data: {
          stateVariableId: "missing-state-id",
        },
      },
    });

    workflowStatePointer.write(writer);

    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });
});
