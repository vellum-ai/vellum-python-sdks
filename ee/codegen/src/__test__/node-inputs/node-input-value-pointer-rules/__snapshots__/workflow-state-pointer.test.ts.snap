// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WorkflowStatePointer > should generate correct Python code 1`] = `
"State.state_1
"
`;

exports[`WorkflowStatePointer > should handle when it's referencing a state variable that no longer exists 1`] = `
"None
"
`;
